keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,294 @@
+    * Only `MaxCSIVolumeCountPred` predicate enabled.
+  * API objects:
+    * Both CSINode and Node containing `status.allocatable` for a volume plugin (to simulate kubelet during deprecation period).
+    * Only CSINode containing `status.allocatable` for a volume plugin (to simulate kubelet after deprecation period).
+    * Only Node containing `status.allocatable` for a volume plugin (to simulate old kubelet).
+  * Test results should be ideally the same as before the KEP.
+    * Any deviation needs to be approved by sig-scheduling.
+
+* Run e2e tests and kubelet version skew tests to check that scheduler picks the right values from CSINode or Node.
+
+* Add e2e test that runs pods with both in-tree volumes and CSI driver for the same storage backend and check that they share the same volume limits.
+
+### Graduation Criteria
+
+##### Alpha -> Beta Graduation
+
+N/A (`AttachVolumeLimit` feature is already beta).
+
+##### Beta -> GA Graduation
+
+It must graduate together with CSI migration.
+
+### Upgrade / Downgrade / Version Skew Strategy
+
+During upgrade, downgrade or version skew, kubelet may be older that scheduler. Kubelet will not fill `CSINode.status` with volume limits and it will fill volume limits into `Node.status`. Scheduler must fall back to `Node.status` when `CSINode` is not available or its `status` does not contain a volume plugin / CSI driver.
+
+#### Interaction with CSI migration
+
+In-tree volume plugins are being migrated to CSI in a [separate KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190129-csi-migration.md). This section covers possible situations:
+
+* CSI migration feature is `off` for a volume plugin and CSI driver is not installed on a node:
+  * In-tree volume plugin is used for in-tree PVs.
+  * Kubelet gets volume limits from the plugin and creates `CSINode` during node registration. Scheduler and kubelet still needs to translate in-tree plugin name to CSI driver name to get the right `name` for `CSINode.status.allocatable`.
+
+* CSI migration feature is `off` for a volume plugin and CSI driver (for the same storage backend) is installed on a node:
+  * In-tree volume plugin is used for in-tree PVs.
+  * CSI driver is used for CSI PVs.
+  * Kubelet gets volume limits from the plugin and creates `CSINode` during node registration.
+  * When CSI driver is registered, kubelet gets its volume limit through CSI and updates `CSINode`, potentially overwriting in-tree plugin limit.
+
+* CSI migration feature is `on` for a volume plugin and there is no CSI driver installed:
+  * In-tree volume plugin is "off".
+  * CSI driver is used both for in-tree and CSI PVs (if the driver is installed).
+  * Kubelet creates `CSINode` during node registration with no limit for the volume plugin / CSI driver.
+  * When CSI driver is registered, kubelet gets its volume limit through CSI and updates `CSINode` with the new limit.
```

Comment:
csi driver wont be registered if no driver is installed.

Reply:
oh I think this step should be after the ` During the period when no CSI driver is registered and CSINode exists` step

Reply:
I reordered the bullets.